sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -465,7 +475,7 @@ abstract class BinaryComparison extends BinaryOperator with Predicate {
     }
   }
 
-  protected lazy val ordering = TypeUtils.getInterpretedOrdering(left.dataType)
+  protected lazy val ordering: Ordering[Any] = TypeUtils.getInterpretedOrdering(left.dataType)
```

Comment:
The acceptable types in `TypeUtils.getInterpretedOrdering` are less than `RowOrdering.isOrderable`. It only accepts `AtomicType`, `ArrayType` and `StructType`.

`NullType`, `UserDefinedType` can cause problems.